### PR TITLE
Improvements to save_as_xml function

### PR DESCRIPTION
**Pull request by davidmurray**
Somewhat related to issue #915.

Commits are quite self-explanatory, but this fixes a few things:
1. A bug where non-existent tags would get written to the XML file when they shouldn't.
2. Adds a way to specify the OSM version to write in the header
3. Adds a way to specify how many decimals to keep when writing lat-long coordinates
4. Sets the encoding to UTF-8. This is useful for areas that have non-ascii characters in the name (for example, in Québec)

Thanks

### CODE REVIEW

**Comment by davidmurray** [comment]
Why don't the CI tests just lint the code directly instead of failing?

File: osmnx/osm_xml.py [hunk 1]
```diff
@@ -163,6 +165,10 @@ def save_graph_xml(
         this method to aggregate the lengths of the individual
         component edges. Otherwise, the length attribute will simply
         reflect the length of the first edge associated with the way.
+    osm_version : int
+        OpenStreetMap data version to write in the XML file header. Default 0.6.
+    coordinate_precision : int
+        Number of decimal places to keep when writing latitude and longitude values. Default 7.
```
Comment (gboeing) [inline osmnx/osm_xml.py:171]: Can you ensure any comment lines are <= 79 characters long?

File: osmnx/osm_xml.py [hunk 2]
```diff
@@ -195,6 +201,8 @@ def save_graph_xml(
 
     # rename columns per osm specification
     gdf_nodes.rename(columns={"x": "lon", "y": "lat"}, inplace=True)
+    gdf_nodes.lon = gdf_nodes.lon.round(coordinate_precision)
+    gdf_nodes.lat = gdf_nodes.lat.round(coordinate_precision)
```
Comment (gboeing) [inline osmnx/osm_xml.py:205]: For Geo/DataFrame columns, please access them via `gdf["col_name"]` rather than `gdf.col_name`.

File: osmnx/osm_xml.py [hunk 3]
```diff
@@ -211,10 +219,6 @@ def save_graph_xml(
         table["changeset"] = "1"
         table["timestamp"] = "2017-01-01T00:00:00Z"
 
-    # convert all datatypes to str
-    gdf_nodes = gdf_nodes.applymap(str)
-    gdf_edges = gdf_edges.applymap(str)
-
```
Comment (gboeing) [inline osmnx/osm_xml.py:217]: Were these lines only necessary for converting `osm_version` to a string? Is there no other edge case where we must convert values to string? I see you added a few other string conversions below: is this more efficient than `applymap(str)`?

**Review by gboeing — CHANGES_REQUESTED** [review]

**Comment by gboeing** [comment]
The test coverage report states:
```
codecov/patch — 92.85% of diff hit (target 97.85%) 
```
When you make the requested changes to the PR, would you check the test coverage?

File: osmnx/osm_xml.py (hunk 3 above)
Comment (davidmurray) [reply osmnx/osm_xml.py:217]: I think they were necessary because `xml.etree` expects strings only. The API allows you to input int and float types, but when you actually try to write the XML file, it fails saying that types must all be strings. Thus, everything needs to be a string before passing to `xml.etree`, but we need the original types to check for null / NaN.

From what I can see, these changes do not break any other edge case based on the testing I've done locally.

**Comment by davidmurray** [comment]
> The test coverage report states:
> 
> ```
> codecov/patch — 92.85% of diff hit (target 97.85%) 
> ```
> 
> When you make the requested changes to the PR, would you check the test coverage?

@gboeing I'm not sure that I understand how the tests work.
For example, these lines in the tests:
https://github.com/gboeing/osmnx/blob/5b87c2b14d78425e9d0fe2d485b8a4f5c75384b3/tests/test_osmnx.py#L171-L172

Unless I am mistaken, this does not test if the output is valid. It merely ensures that the code hasn't produced any exceptions. A real test suite would look at the written data and ensure that it is valid OSM xml. Is this correct?
In this case, how would you suggest that I modify the code to improve the test coverage?

Thank you very much

**Comment by gboeing** [comment]
> Unless I am mistaken, this does not test if the output is valid. It merely ensures that the code hasn't produced any exceptions. A real test suite would look at the written data and ensure that it is valid OSM xml. Is this correct?
In this case, how would you suggest that I modify the code to improve the test coverage?

1) yes it looks like the XML tests are just checking for exceptions. At a minimum, could you ensure that the lines in your diff are hit by the tests and no exceptions result?

2) ideally yes it would be great to validate the XML output. I do not work with OSM XML files myself so my expertise there is limited. Are you aware of any validator tools?

**Comment by gboeing** [comment]
Let me know if this PR is still under development and should remain open.

**Comment by davidmurray** [comment]
> Let me know if this PR is still under development and should remain open.

Hi, yeah I haven't had time to finish it up but I'll get it done soon.

**Comment by gboeing** [comment]
I've been going through and closing out stale PRs and issues. Let me know if this remains active.

**Comment by davidmurray** [comment]
@gboeing Sorry, don't think I'll have time to fix the unit tests. However, the improvements in this PR are valuable and I think they should be merged (or considered) to be merged in the main tree. Cheers

**Comment by gboeing** [comment]
> this does not test if the output is valid. It merely ensures that the code hasn't produced any exceptions. A real test suite would look at the written data and ensure that it is valid OSM xml

#1135 adds an OSM XML schema XSD file and now validates saved OSM XML files against it in the test suite.
